classes/ActionScheduler_ListTable.php

```diff
@@ -0,0 +1,385 @@
+		}
+
+		return __( 'Pending', 'action-scheduler' );
+	}
+
+	/**
+	 * Bulk delete
+	 *
+	 * Deletes actions based on their ID. This is the handler for the bulk delete. It assumes the data
+	 * properly validated by the callee and it will delete the actions without any extra validation.
+	 *
+	 * @param array $ids
+	 *
+	 * @return void
+	 */
+	protected function bulk_delete( array $ids, $ids_sql ) {
+		foreach ( $ids as $id ) {
+			$this->stores->action->delete_action( $id );
+		}
+	}
+
+	/**
+	 * Implements the logic behind running an action. PP_Table_List validates the request and their
+	 * parameters are valid.
+	 */
+	protected function row_action_run( $action_id ) {
+		try {
+			ActionScheduler::runner()->process_action( $action_id );
+			$success = 1;
+		} catch ( Exception $e ) {
+			$success = 0;
+		}
+
+		set_transient( 'actionscheduler_admin_executed', compact( 'action_id', 'success' ), 30 );
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function prepare_items() {
+		$this->process_bulk_action();
+
+		$this->process_row_actions();
+
+		if ( ! empty( $_REQUEST['_wp_http_referer'] ) ) {
+			// _wp_http_referer is used only on bulk actions, we remove it to keep the $_GET shorter
+			wp_redirect( remove_query_arg( array( '_wp_http_referer', '_wpnonce' ), wp_unslash( $_SERVER['REQUEST_URI'] ) ) );
+			exit;
+		}
+
+		$this->prepare_column_headers();
+
+		$per_page = $this->get_items_query_limit();
+		$query = array(
+			'per_page' => $per_page,
+			'offset'   => $this->get_items_query_offset(),
+			'status'   => $this->get_request_status(),
+			'orderby'  => 'modified',
+			'order'    => 'ASC',
+		);
+
+		$this->items = array();
+
+		$total_items = $this->stores->action->query_actions_count( $query );
+
+		foreach ( $this->stores->action->query_actions( $query ) as $id ) {
+			$item = $this->stores->action->fetch_action( $id );
+			$this->items[ $id ] = array(
+				'ID'     => $id,
+				'hook'   => $item->get_hook(),
+				'status' => $this->get_status( $item ),
+				'args'   => $item->get_args(),
+				'group'  => $item->get_group(),
+				'recurrence' => $this->get_recurrence( $item ),
+				'scheduled'  => $item->get_schedule(),
+				'claim_id'   => '',
+				'comments'   => $this->stores->log->get_logs( $id ),
+			);
+		}
+
+		$this->set_pagination_args( array(
+			'total_items' => $total_items,
+			'per_page'    => $per_page,
+			'total_pages' => ceil( $total_items / $per_page ),
+		) );
+	}
+
+	/**
+	 * This function is used to filter the actions by a status. It will return the status based on a 
+	 * a GET parameter or it will filter by 'pending' by default. 
+	 * We filter by status because if we do not filter we have NullActions that are not rendered at the 
+	 * moment.
+	 *
+	 * @return string
+	 */
+	public function get_request_status() {
+		$statuses = array(
+			ActionScheduler_Store::STATUS_PENDING,
+			ActionScheduler_Store::STATUS_COMPLETE,
+			ActionScheduler_Store::STATUS_FAILED,
```

Comment:
What about `ActionScheduler_Store::STATUS_CANCELED` and `ActionScheduler_Store::STATUS_RUNNING`?

Reply:
> Do you want me to fix it @thenbrent ? 

Yes. But make sure your fix is using the least possible code. Don't just add all those methods `ActionScheduler_NullAction`. That is most likely not required. Perhaps instead create a new `ActionScheduler_CancelledAction` which extends `ActionScheduler_Action` or similar.

Reply:
Will do it now.

Reply:
I should also say be sure to double and triple check all occasions where a cancelled action is instantiated to see what other implications the above suggestion will have. Changing the way cancelled actions are instantiated could have terrible consequences, like duplicating actions.

There are likely other possible solutions with lower impact than that suggest, like manually instantiating an `ActionScheduler_Action` instance for cancelled actions in this code to avoid impacting any other code. Please explore that and any other possible options you come up with too.

Reply:
> There are likely other possible solutions with lower impact than that suggest, like manually instantiating an ActionScheduler_Action instance for cancelled actions in this code to avoid impacting any other code. Please explore that and any other possible options you come up with too.

This is a hack, but I agree, it may has with lower impact. However I'm think I should change https://github.com/Prospress/action-scheduler/blob/master/classes/ActionScheduler_wpPostStore.php#L96-L97 to instantiate a 'CancelledAction' object if the `post_status` is `'trash'`. It does not seem that error prone. I haven't look at the custom tables repo though.

Reply:
> It does not seem that error prone. I haven't look at the custom tables repo though.

It's not the custom tables repo I'm worried about. It's the queue runner.

> I'm think I should change

I can't see why that wouldn't return a valid action for `ActionScheduler_Store::STATUS_CANCELED`?

I just had a look at `make_action_from_post()`, and I can't see why that would return an instance of `ActionScheduler_NullAction` either. It should return an instance of `ActionScheduler_FinishedAction`.